benchmarks/tensorexpr/microbenchmarks.py

```diff
@@ -0,0 +1,256 @@
+    ("asin", torch.asin),
+    ("acos", torch.acos),
+    ("atan", torch.atan),
+    ("sinh", torch.sinh),
+    ("cosh", torch.cosh),
+    ("tanh", torch.tanh),
+    ("sigmoid", torch.sigmoid),
+    ("exp", torch.exp),
+    ("expm1", torch.expm1),
+    ("expm1", torch.expm1),
+    ("abs", torch.abs),
+    ("log", torch.log),
+    ("fast_log", torch.log),
+    ("log2", torch.log2),
+    ("log10", torch.log10),
+    ("log1p", torch.log1p),
+    ("erf", torch.erf),
+    ("erfc", torch.erfc),
+    ("sqrt", torch.sqrt),
+    ("rsqrt", torch.rsqrt),
+    ("ceil", torch.ceil),
+    ("floor", torch.floor),
+    ("round", torch.round),
+    ("trunc", torch.trunc),
+    ("lgamma", torch.lgamma),
+    # ("frac", torch.frac), # seems unimplemented
+    # ("isnan", torch.isnan), # no out variant
+]
+
+def gen_unary_nnc_fun(nnc_name):
+    def nnc_fun(A, B):
+        def compute(i, j):
+            return getattr(A.load([i, j]), nnc_name)()
+        return compute
+    return nnc_fun
+
+def gen_unary_torch_fun(torch_op):
+    def torch_fun(a, b, out):
+        def fun():
+            torch_op(a, out=out)
+        return fun
+    return torch_fun
+
+
+def gen_binary_nnc_fun(fn):
+    def nnc_fun(A, B):
+        def compute(i, j):
+            return fn(A.load([i, j]), B.load([i, j]))
+        return compute
+    return nnc_fun
+
+def gen_binary_torch_fun(fn):
+    def pt_fun(a, b, out):
+        def fun():
+            return fn(a, b, out=out)
+        return fun
+    return pt_fun
+
+def gen_int_comparison_tensors(N, M):
+    return (torch.randint(0, 3, (N, M)), torch.randint(0, 3, (N, M)), torch.empty((N, M), dtype=torch.bool))
+
+def gen_float_comparison_tensors(N, M):
+    return (torch.rand(N, M), torch.rand(N, M), torch.empty((N, M), dtype=torch.bool))
+
+
+te_bool = te.Dtype.Bool
+binary_ops = [
+    ('add', (lambda a, b: a + b), torch.add),
+    ('mul', (lambda a, b: a * b), torch.mul),
+    ('sub', (lambda a, b: a - b), torch.sub),
+    ('div', (lambda a, b: a / b), torch.div),
+    ('eq', (lambda a, b: te.Cast.make(te_bool, a == b)), torch.eq, gen_int_comparison_tensors),
+    ('gt', (lambda a, b: te.Cast.make(te_bool, a > b)), torch.gt, gen_float_comparison_tensors),
+    ('lt', (lambda a, b: te.Cast.make(te_bool, a < b)), torch.lt, gen_float_comparison_tensors),
+    ('gte', (lambda a, b: te.Cast.make(te_bool, a >= b)), torch.greater_equal, gen_float_comparison_tensors),
+    ('lte', (lambda a, b: te.Cast.make(te_bool, a <= b)), torch.less_equal, gen_float_comparison_tensors),
+    # ('neq', (lambda a, b: a != b), None)), # no one-op equivalent
+    # ('&', (lambda a, b: a & b), torch.bitwise_and), # requires more work to test
+]
+
+
+def nnc_relu(A, B):
+    def f(i, j):
+        return torch._C.te.ifThenElse(A.load([i, j]) < torch._C.te.ExprHandle.float(0),
+                                      torch._C.te.ExprHandle.float(0), A.load([i, j]))
+    return f
+
+custom_ops = [
+    ('relu', nnc_relu, lambda a, b, c: c.copy_(a).relu_()),
```

Comment:
I don't think this relu is totally fair to pytorch.  It involves a read-modify-write of `a` for `relu_()` and then `read a, write to c` for `copy_`.  That's 1.5 as much memory traffic as you need to do `c = a.relu()`, which is what NNC is essentially doing.

Reply:
You're right - I wanted to do this to avoid a memory allocation, but just directly returning `a.relu()` seems to be faster.

Reply:
NNC still seems to be about 20G% faster or so, but definitely not nearly as good as the previous `70%` :(